datafusion/physical-plan/src/aggregates/group_values/multi_group_by/mod.rs

```diff
@@ -491,74 +528,168 @@ impl<const STREAMING: bool> GroupValuesColumn<STREAMING> {
-                // for hasher function, use precomputed hash value
-                self.map.insert_accounted(
-                    (target_hash, group_index_view),
-                    |(hash, _)| *hash,
-                    &mut self.map_size,
-                );
-
-                // Add row index to `vectorized_append_row_indices`
-                self.vectorized_operation_buffers
-                    .append_row_indices
-                    .push(row);
-
-                // Set group index to row in `groups`
-                groups[row] = current_group_idx;
-
-                group_values_len += 1;
-                continue;
-            };
+            for (row, &target_hash) in batch_hashes_iter {
+                // prefetch next item
+                for i in 1..=PREFETCH {
+                    if READ {
+                        self.map.prefetch_read::<LOCALITY>(batch_hashes[row + i]);
+                    } else {
+                        self.map.prefetch_write::<LOCALITY>(batch_hashes[row + i]);
+                    }
+                }
+                self.insert_entry(groups, &mut group_values_len, row, target_hash);
+            }
+            for index in batch_hashes.len().saturating_sub(PREFETCH)..batch_hashes.len() {
+                self.insert_entry(groups, &mut group_values_len, index, batch_hashes[index]);
+            }
+        } else {
+            let mut group_values_len = self.group_values[0].len();
 
-            // 2. bucket found
-            // Check if the `group index view` is `inlined` or `non_inlined`
-            if group_index_view.is_non_inlined() {
-                // Non-inlined case, the value of view is offset in `group_index_lists`.
-                // We use it to get `group_index_list`, and add related `rows` and `group_indices`
-                // into `vectorized_equal_to_row_indices` and `vectorized_equal_to_group_indices`.
-                let list_offset = group_index_view.value() as usize;
-                let group_index_list = &self.group_index_lists[list_offset];
-
-                self.vectorized_operation_buffers
-                    .equal_to_group_indices
-                    .extend_from_slice(group_index_list);
-                self.vectorized_operation_buffers
-                    .equal_to_row_indices
-                    .extend(std::iter::repeat_n(row, group_index_list.len()));
-            } else {
-                let group_index = group_index_view.value() as usize;
-                self.vectorized_operation_buffers
-                    .equal_to_row_indices
-                    .push(row);
-                self.vectorized_operation_buffers
-                    .equal_to_group_indices
-                    .push(group_index);
+            for (row, &target_hash) in batch_hashes.iter().enumerate() {
+                self.insert_entry(groups, &mut group_values_len, row, target_hash);
             }
         }
     }
 
+    #[inline(always)]
```

Comment:
Inline always to make sure my change for extracting to a function is not affecting the prev case